rllib/models/specs/specs_base.py

```diff
@@ -0,0 +1,209 @@
+        """
+        sampled_shape = self._sample_shape()
+        return self._sample(sampled_shape, fill_value)
+
+    @abc.abstractmethod
+    def _sample(
+        self, shape: Tuple[int], fill_value: Union[float, int] = 0
+    ) -> TensorType:
+        """Returns a sample tensor with the given shape.
+
+        Args:
+            shape: The shape of the tensor to be sampled.
+            fill_value: The value to fill the tensor with.
+
+        Returns:
+            A tensor with the specified value that matches the specs.
+        """
+        raise NotImplementedError
+
+    def _sample_shape(self) -> Tuple[int]:
+        """Converts the expected shape to a sample shape by replacing TBD dimensions
+        with size of 1."""
+        sampled_shape = tuple()
+        for d in self._expected_shape:
+            if isinstance(d, int):
+                sampled_shape += (d,)
+            else:
+                sampled_shape += (1,)
+        return sampled_shape
+
+    def _parse_expected_shape(self, shape: str, shape_vals: Dict[str, int]) -> tuple:
+        """Converts the input shape to a tuple of integers and strings."""
+        expected_shape = tuple()
+
+        # check the validity of shape_vals and get a list of dimension names
+        d_names = self._validate_shape_vals(shape, shape_vals)
+
+        for d in d_names:
+            d_value = shape_vals.get(d, None)
+            if d_value is None:
+                d_value = d
+            expected_shape += (d if d_value is None else d_value,)
+
+        return expected_shape
+
+    def _validate_shape_vals(self, shape: str, shape_vals: Dict[str, int]) -> List[str]:
+        """Checks if the shape_vals is valid and returns a list of dimension names.
+
+        Valid means that shape consist of unique dimension names and shape_vals only
+        consists of keys that are in shape. Also shape_vals can only contain postive
+        integers.
+        """
+        d_names = shape.split(" ")
+        d_names_set = set(d_names)
+        if len(d_names_set) != len(d_names):
+            raise ValueError(_INVALID_INPUT_DUP_DIM.format(",".join(d_names)))
+
+        for d_name in shape_vals:
+            if d_name not in d_names_set:
+                raise ValueError(
+                    _INVALID_INPUT_Unknown_DIM.format(d_name, ",".join(d_names))
+                )
+
+            d_value = shape_vals.get(d_name, None)
+            if d_value is not None:
+                if not isinstance(d_value, int):
+                    raise ValueError(
+                        _INVALID_INPUT_NONINT_DIM.format(d_name, shape, type(d_value))
+                    )
+                if d_value <= 0:
+                    raise ValueError(
+                        _INVALID_INPUT_NON_POSITIVE.format(d_name, ",".join(d_names))
+                    )
+
+        return d_names
+
+    def __repr__(self) -> str:
+        return f"TensorSpec(shape={tuple(self.shape)}, dtype={self.dtype})"
```

Comment:
itd be cool if we could have one of those compressed print outs of the data inside of the tensor, kinda like in numpy or torch.

Reply:
This is not data. This is Tensor spec which doesn't own any data.